staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/envelope.go

```diff
@@ -43,6 +44,52 @@ type Service interface {
 	Encrypt(data []byte) ([]byte, error)
 }
 
+type dek struct {
+	plaintext  []byte
+	ciphertext []byte
+	rotateAt   time.Time
+	mux        sync.RWMutex
+}
+
+func (d *dek) shouldRotate() bool {
+	d.mux.RLock()
+	defer d.mux.RUnlock()
+	return time.Now().After(d.rotateAt)
+}
+
+func (d *dek) rotate(service Service, ttl time.Duration) error {
+	var err error
+	d.plaintext, err = generateKey(32)
+	if err != nil {
+		return err
+	}
+
+	d.mux.Lock()
+	defer d.mux.Unlock()
```

Comment:
check `d.rotateAt` here again, in case another goroutine called `rotate` in between this goroutine's `shouldRotate` and `rotate`